cpp/src/arrow/engine/substrait/expression_internal.cc

```diff
@@ -0,0 +1,427 @@
+
+    case st::Expression::Literal::kFp32:
+      return Datum(lit.fp32());
+    case st::Expression::Literal::kFp64:
+      return Datum(lit.fp64());
+
+    case st::Expression::Literal::kString:
+      return Datum(lit.string());
+    case st::Expression::Literal::kBinary:
+      return Datum(std::make_shared<BinaryScalar>(Buffer::FromString(lit.binary())));
+
+    case st::Expression::Literal::kTimestamp:
+      return Datum(std::make_shared<TimestampScalar>(
+          static_cast<int64_t>(lit.timestamp()), TimeUnit::MICRO));
+
+    case st::Expression::Literal::kTimestampTz:
+      return Datum(std::make_shared<TimestampScalar>(
+          static_cast<int64_t>(lit.timestamp_tz()), TimeUnit::MICRO,
+          TimestampTzTimezoneString()));
+
+    case st::Expression::Literal::kDate:
+      return Datum(std::make_shared<Date32Scalar>(static_cast<int32_t>(lit.date())));
+    case st::Expression::Literal::kTime:
+      return Datum(std::make_shared<Time64Scalar>(static_cast<int64_t>(lit.time()),
+                                                  TimeUnit::MICRO));
+
+    case st::Expression::Literal::kIntervalYearToMonth:
+    case st::Expression::Literal::kIntervalDayToSecond:
+      break;
+
+    case st::Expression::Literal::kUuid:
+      return Datum(std::make_shared<ExtensionScalar>(
+          FixedSizeBinaryScalarFromBytes(lit.uuid()), uuid()));
+
+    case st::Expression::Literal::kFixedChar:
+      return Datum(std::make_shared<ExtensionScalar>(
+          FixedSizeBinaryScalarFromBytes(lit.fixed_char()),
+          fixed_char(static_cast<int32_t>(lit.fixed_char().size()))));
+
+    case st::Expression::Literal::kVarChar:
+      // FIXME
+      // There's no way to determine VarChar.length from the literal
+      break;
+
+    case st::Expression::Literal::kFixedBinary:
+      return Datum(FixedSizeBinaryScalarFromBytes(lit.fixed_char()));
+
+    case st::Expression::Literal::kDecimal:
+      if (lit.decimal().size() != sizeof(Decimal128)) {
+        return Status::Invalid("Decimal literal had ", lit.decimal().size(),
+                               " bytes (expected ", sizeof(Decimal128), ")");
+      }
+
+      // FIXME
+      // It's not clear how these bytes should be interpreted...
+      // Furthermore, there's no way to determine scale or precision
+      break;
+
+    case st::Expression::Literal::kStruct: {
+      const auto& struct_ = lit.struct_();
+
+      ScalarVector fields(struct_.fields_size());
+      std::vector<std::string> field_names(fields.size(), "");
+      for (size_t i = 0; i < fields.size(); ++i) {
+        ARROW_ASSIGN_OR_RAISE(auto field, FromProto(struct_.fields(i)));
+        DCHECK(field.is_scalar());
+        fields.push_back(field.scalar());
+      }
+      ARROW_ASSIGN_OR_RAISE(
+          auto scalar, StructScalar::Make(std::move(fields), std::move(field_names)));
+      return Datum(std::move(scalar));
+    }
+
+      // case st::Expression::Literal::kNamedStruct:
```

Comment:
This comment can be removed, because `NamedStruct` isn't part of the type union. It's only used in the `base_schema` field of `ReadRel`.